storage-proofs/src/layered_drgporep.rs

```diff
@@ -363,110 +365,89 @@ pub trait Layers {
         let mut taus = Vec::with_capacity(layers);
         let mut auxs: Vec<Tree<Self::Hasher>> = Vec::with_capacity(layers);
 
-        // We need to create a scope which encloses all the work, spawning threads
-        // for merkle tree generation and sending the results back to a channel.
-        // The received results need to be sorted by layer because ordering of the completed results
-        // is not guaranteed. Misordered results will be seen in practice when trees are small.
-
-        // The outer scope ensure that `tx` is dropped and closed before we read from `outer_rx`.
-        // Otherwise, the read loop will block forever waiting for more input.
-        let outer_rx = {
-            let (tx, rx) = channel();
-
-            let errf = |e| {
-                let err_string = format!("{:?}", e);
-                error!(
-                    "MerkleTreeGenerationError: {} - {:?}",
-                    &err_string,
-                    failure::Backtrace::new()
-                );
-                Error::MerkleTreeGenerationError(err_string)
-            };
-
-            let _ = thread::scope(|scope| -> Result<()> {
-                let mut threads = Vec::with_capacity(layers + 1);
-                (0..=layers).fold(graph.clone(), |current_graph, layer| {
-                    let leafs = data.len() / NODE_SIZE;
-                    assert_eq!(data.len() % NODE_SIZE, 0);
-                    let pow = next_pow2(leafs);
-                    // FIXME: Who's actually responsible for ensuring power of 2
-                    //  sector sizes?
-
-                    let mut leaves_store = MerkleStore::new(pow);
-
-                    populate_leaves::<
-                        _,
-                        <Self::Hasher as Hasher>::Function,
-                        _,
-                        std::iter::Map<_, _>,
-                    >(&mut leaves_store, (0..leafs).map(|i| {
+        let errf = |e| {
+            let err_string = format!("{:?}", e);
+            error!(
+                "MerkleTreeGenerationError: {} - {:?}",
+                &err_string,
+                failure::Backtrace::new()
+            );
+            Error::MerkleTreeGenerationError(err_string)
+        };
+
+        let merkle_trees = Arc::new(Mutex::new(vec![None; layers + 1]));
```

Comment:
Is there a nicer way to have a `Vector` with already all the positions available for an our of order write to it? Should we use references instead?